design/cluster-api-provider-metal3/node_reuse.md

```diff
@@ -0,0 +1,225 @@
+
+#### Node Reuse Approach
+
+We should perform two steps to re-use the same pool of hosts.
+
+1. During host deprovisioning, set the `infrastructure.cluster.x-k8s.io/node-reuse`
+  label;
+2. During next provisioning, filter out hosts with the label.
+
+The actual implementation will be done within the CAPM3 Machine controller.
+
+Step1 can be done as follows:
+
+- CAPM3 controller sets the `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1`
+  label on the hosts belonging to the same MD/KCP.
+
+<p align="center">
+    <img src="../images/node_reuse_deprovisioning.png"/>
+    <br>
+    <em>diagram-1: Example flow of deprovisioning process</em>
+</p>
+
+Step2 can be done as follows:
+
+- CAPM3 controller filters host labelled with
+  `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1` label.
+
+  Next:
+  - If host is found in `Ready` state:
+    - Pick that host for newly created M3M;
+    - Once it is picked up, remove the whole label
+    (`infrastructure.cluster.x-k8s.io/node-reuse: md-pool1`)
+    from the host.
+  - If host is found in `Deprovisioning` state:
+    - Requeue until that host becomes `Ready`;
+  - If no host is found, while it should be (i.e for some reason host
+    is not in the cluster anymore):
+    - Fall back to the current flow, which selects host randomly.
+
+<p align="center">
+    <img src="../images/node_reuse_provisioning.png"/>
+    <br>
+    <em>diagram-2: Example flow of provisioning process</em>
+</p>
+
+**What if there are two parallel MD/KCP experiencing upgrade?**
+
+  Two MachineDeployments(md-pool1 and md-pool2) with nodeReuse set to True & False
+  respectively start upgrade at the same time. During provisioning, CAPM3 Machine
+  controller tries to find hosts with
+  `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1` label for md-pool1, and
+  `infrastructure.cluster.x-k8s.io/node-reuse: md-pool2` label for md-pool2.
+  Based on the logic on diagram-2 above, for md-pool1, machine controller finds
+  hosts with the md-pool1 label and picks them for provisioning. For md-pool2,
+  controller searches for hosts with md-pool2 label and fails to find any
+  (because label wasn't set during deprovisioning due to disabled NodeReuse). As
+  such, controller falls back to select random hosts.
+
+**What if we found N-1 hosts with label, while we expect N hosts?**
+  Then we select any other hosts, even though it has no label we expect.
+
+**What if user changes from node reuse to no node reuse in the middle of upgrade?**
+
+  In the middle of the upgrade process, when hosts (already labelled during
+  deprovisioning) are provisioning, user disables nodeReuse. Based on the
+  diagram-2 above, CAPM3 controller always picks labelled hosts (when found)
+  regardless of nodeReuse feature is disabled/enabled. The reason is, we want to
+  ensure that the controller doesn't miss labelled hosts. In short, labelled
+  hosts always take the presedence over unlabelled hosts during selection.
+
```

Comment:
Controller will always try to select hosts with the label, no matter if node reuse is enabled/disabled. As such, during the next upgrade/re-provisioning, if there is a host with the label, then yes controller will pick that host (even though it was labelled before the last deprovisioning).
Why do you think that host won't be selected? Do you mean in some corner cases, due to some issues that while selection, that host with the label wasn't available (for some reason), but became available after controller has picked another host?

Reply:
Yes, I was guessing if there was any corner case such that, and if it could had any drawbacks, as it means that the reuse procedure may have not succeeded (since an host marked for reuse wasn't picked up). I guess that probably it would be sufficient to log such scenario (ie, when nodeReuse is set but no hosts were found with the `infrastructure.cluster.x-k8s.io/node-reuse` during the provisioning)

Reply:
@andfasano sure thing, we will tackle the logging of such scenario in the implementation.